pkg/runtime/hotreload/loader/operator/generic.go

```diff
@@ -0,0 +1,176 @@
+/*
+Copyright 2023 The Dapr Authors
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package operator
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"sync"
+	"sync/atomic"
+
+	"github.com/cenkalti/backoff/v4"
+
+	operatorpb "github.com/dapr/dapr/pkg/proto/operator/v1"
+	"github.com/dapr/dapr/pkg/runtime/hotreload/differ"
+	"github.com/dapr/dapr/pkg/runtime/hotreload/loader"
+	"github.com/dapr/dapr/pkg/runtime/hotreload/loader/store"
+)
+
+type generic[T differ.Resource] struct {
+	opClient  operatorpb.OperatorClient
+	podName   string
+	namespace string
+
+	streamer streamer[T]
+	store    store.Store[T]
+
+	wg      sync.WaitGroup
+	closeCh chan struct{}
+	closed  atomic.Bool
+}
+
+type streamer[T differ.Resource] interface {
+	list(ctx context.Context, opclient operatorpb.OperatorClient, ns, podName string) ([][]byte, error)
+	close() error
+	recv() (*loader.Event[T], error)
+	establish(context.Context, operatorpb.OperatorClient, string, string) error
+}
+
+func newGeneric[T differ.Resource](opts Options, store store.Store[T], streamer streamer[T]) *generic[T] {
+	return &generic[T]{
+		opClient:  opts.OperatorClient,
+		podName:   opts.PodName,
+		namespace: opts.Namespace,
+		streamer:  streamer,
+		store:     store,
+		closeCh:   make(chan struct{}),
+	}
+}
+
+func (g *generic[T]) List(ctx context.Context) (*differ.LocalRemoteResources[T], error) {
+	var resp [][]byte
+	err := backoffFn(ctx, func(ctx context.Context) error {
+		var berr error
+		resp, berr = g.streamer.list(ctx, g.opClient, g.namespace, g.podName)
```

Comment:
This is a side effect and will cause a race condition.

Reply:
Not sure I follow, why is there a race condition here?

Reply:
the callback function runs in a background goroutine, and it's modifying `resp` outside.